File: src/main/java/lestermartin/hadoop/exploration/opengeorgia/SalaryReportBuilder.java
package lestermartin.hadoop.exploration.opengeorgia;


/**
 * class is NOT thread safe!!
 */
public class SalaryReportBuilder {

    private String workingText;

    /**
     * ON PURPOSE I'm NOT using something more sophisticated such as Apache's Commons...
     */
    public SalaryReport buildFromInputCsvFormat(String textAsCsv) {
        workingText = textAsCsv;
        SalaryReport salaryReport = new SalaryReport();

        salaryReport.setName( parseNextValueAsString() );
        salaryReport.setTitle( parseNextValueAsString() );

        salaryReport.setSalary( parseNextValueAsFloat() );
        salaryReport.setTravel( parseNextValueAsFloat() );

        salaryReport.setOrgType( parseNextValueAsString() );
        salaryReport.setOrg( parseNextValueAsString() );

        salaryReport.setYear( Integer.valueOf( workingText.substring(0) ).intValue() );

        //sanity check
        if( salaryReport.getYear() > 2050  ||  salaryReport.getYear() < 2000 ) {
            throw new IllegalStateException("unable to correct parse string '" + textAsCsv +
                    "' -- incorrectly parsed into '" + salaryReport.toString() + "'");
        }

        return salaryReport;
    }


    private String parseNextValueAsString() {
        String nextValue = null;
        //field might, or might not, start with double-quotes
        if(workingText.startsWith("\"")) {
            int endOfValue = workingText.indexOf("\"", 1);
            nextValue = workingText.substring(1, endOfValue);
            workingText = workingText.substring(endOfValue + 2);
        } else {
            int endOfValue = workingText.indexOf(",", 0);
            nextValue = workingText.substring(0, endOfValue);
            workingText = workingText.substring(endOfValue + 1);
        }
        return nextValue;
    }

    private float parseNextValueAsFloat() {
        String floatAsString = parseNextValueAsString();
        floatAsString = floatAsString.replace("$", "");
        floatAsString = floatAsString.replace(",", "");
        floatAsString = floatAsString.replace(" ", "");

        return Float.valueOf(floatAsString).floatValue();
    }

}

File: src/main/csharp/opengeorgia/SalaryReportBuilder.cs
using System;

namespace OpenGeorgia
{
    public class SalaryReportBuilder
    {
        private string workingText;

        public SalaryReport buildFromInputCsvFormat(string textAsCsv)
        {
            workingText = textAsCsv;
            SalaryReport salaryReport = new SalaryReport();

            salaryReport.Name = parseNextValueAsString();
            salaryReport.Title = parseNextValueAsString();

            salaryReport.Salary = parseNextValueAsFloat();

            salaryReport.Travel = parseNextValueAsFloat();

            salaryReport.OrgType = parseNextValueAsString();
            salaryReport.Org = parseNextValueAsString();

            salaryReport.Year = int.Parse(workingText.Substring(0));

            //sanity check!!
            if (salaryReport.Year > 2050 || salaryReport.Year < 2000)
            {
                throw new ArgumentException("unable to correctly parse string '" + textAsCsv + "'");
            }

            return salaryReport;
        }

        private String parseNextValueAsString()
        {
            string nextValue = null;
            //field, might, or might not, start with double-quotes
            if (workingText.StartsWith("\""))
            {
                int endOfValue = workingText.IndexOf("\"", 1);
                nextValue = workingText.Substring(1, endOfValue-1);
                workingText = workingText.Substring(endOfValue + 2);
            }
            else
            {
                int endOfValue = workingText.IndexOf(",", 0);
                nextValue = workingText.Substring(0, endOfValue);
                workingText = workingText.Substring(endOfValue + 1);
            }
            return nextValue;
        }

        private float parseNextValueAsFloat()
        {
            String floatAsString = parseNextValueAsString();
            floatAsString = floatAsString.Replace("$", "");
            floatAsString = floatAsString.Replace(",", "");
            floatAsString = floatAsString.Replace(" ", "");

            return float.Parse(floatAsString);
        }

    }
}
